packages/flutter/lib/src/widgets/modal_barrier.dart

```diff
@@ -14,6 +14,119 @@ import 'gesture_detector.dart';
 import 'navigator.dart';
 import 'transitions.dart';
 
+enum ClipDirection {
+  left,
+  right,
+  top,
+  bottom,
+  none,
+}
+
+class BarrierClipper extends SingleChildRenderObjectWidget{
+  const BarrierClipper({
+    super.key,
+    super.child,
+    this.clipDirection = ClipDirection.none,
+    this.topLayerSizeNotifier,
+  });
+
+  final ClipDirection clipDirection;
+  final ValueNotifier<Size>? topLayerSizeNotifier;
+
+  @override
+  RenderBarrierClipper createRenderObject(BuildContext context) {
+    return RenderBarrierClipper(clipDirection: clipDirection, topLayerSizeNotifier: topLayerSizeNotifier,);
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, RenderBarrierClipper renderObject) {
+    renderObject
+    ..clipDirection = clipDirection
+    ..topLayerSizeNotifier = topLayerSizeNotifier;
+  }
+}
+
+class RenderBarrierClipper extends RenderProxyBox {
+  RenderBarrierClipper({
+    required ClipDirection clipDirection,
+    ValueNotifier<Size>? topLayerSizeNotifier,
+    RenderBox? child,
+  }) : _clipDirection = clipDirection,
+       _topLayerSizeNotifier = topLayerSizeNotifier,
+       assert(clipDirection != null),
+       super(child);
+
+  ValueNotifier<Size>? _topLayerSizeNotifier;
+  ValueNotifier<Size>? get topLayerSizeNotifier => _topLayerSizeNotifier;
+  set topLayerSizeNotifier (ValueNotifier<Size>? newNotifier) {
+    if (_topLayerSizeNotifier == newNotifier) {
+      return;
+    }
+    _topLayerSizeNotifier = newNotifier;
```

Comment:
Some where in this class should listen to Notifier and markNeedsSemanticsUpdate
likely in attach and detach